KosherCocoa/Library/Core/Solar/KCNOAACalculator.m

```diff
@@ -0,0 +1,450 @@
+    double solarDec = [self getSunDeclination:tnoon];
+    double hourAngle = [self getSunHourAngleAtSunrise:latitude solarDec:solarDec zenith:zenith];
+    
+    double delta = -longitude - toDegrees(hourAngle);
+    double timeDiff = 4 * delta;
+    double timeUTC = 720 + timeDiff - eqTime;
+    
+    // Second pass includes fractional Julian Day in gamma calc
+    double newt = [self getJulianCenturiesFromJulianDay:[self getJulianDayFromJulianCenturies:julianCenturies] + timeUTC/ 1440.0];
+    eqTime = [self getEquationOfTime:newt];
+    solarDec = [self getSunDeclination:newt];
+    hourAngle = [self getSunHourAngleAtSunrise:latitude solarDec:solarDec zenith:zenith];
+    
+    delta = -longitude - toDegrees(hourAngle);
+    timeDiff = 4 * delta;
+    timeUTC = 720 + timeDiff - eqTime;
+    return timeUTC;
+    }
+
+- (double)getSolarNoonUTC:(double)julianCenturies longitude:(double)longitude {
+    // First pass uses approximate solar noon to calculate equation of time
+    double tnoon = [self getJulianCenturiesFromJulianDay:[self getJulianDayFromJulianCenturies:julianCenturies] + longitude / 360.0];
+    double eqTime = [self getEquationOfTime:tnoon];
+    double solNoonUTC = 720 + (longitude * 4) - eqTime; // min
+    
+    double newt = [self getJulianCenturiesFromJulianDay:[self getJulianDayFromJulianCenturies:julianCenturies] - 0.5 + solNoonUTC / 1440.0];
+    eqTime = [self getEquationOfTime:newt];
+    return 720 + (longitude * 4) - eqTime; // min
+}
+
+
+- (double)getSunsetUTC:(double)julianDay latitude:(double)latitude longitude:(double)longitude zenith:(double)zenith {
+    double julianCenturies = [self getJulianCenturiesFromJulianDay:julianDay];
+    
+    // Find the time of solar noon at the location, and use that declination. This is better than start of the
+    // Julian day
+    double noonmin = [self getSolarNoonUTC:julianCenturies longitude:-longitude];
+    double tnoon = [self getJulianCenturiesFromJulianDay:julianDay + noonmin / 1440.0];
+    
+    // First calculates sunrise and approx length of day
+    double eqTime = [self getEquationOfTime:tnoon];
+    double solarDec = [self getSunDeclination:tnoon];
+    double hourAngle = [self getSunHourAngleAtSunset:latitude solarDec:solarDec zenith:zenith];
+    
+    double delta = -longitude - toDegrees(hourAngle);
+    double timeDiff = 4 * delta;
+    double timeUTC = 720 + timeDiff - eqTime;
+    
+    // Second pass includes fractional Julian Day in gamma calc
+    double newt = [self getJulianCenturiesFromJulianDay:[self getJulianDayFromJulianCenturies:julianCenturies] + timeUTC/ 1440.0];
+    eqTime = [self getEquationOfTime:newt];
+    solarDec = [self getSunDeclination:newt];
+    hourAngle = [self getSunHourAngleAtSunset:latitude solarDec:solarDec zenith:zenith];
+    
+    delta = -longitude - toDegrees(hourAngle);
+    timeDiff = 4 * delta;
+    timeUTC = 720 + timeDiff - eqTime;
+    return timeUTC;
+}
+
+
+#pragma mark - Methods from Astronomical Calculator
```

Comment:
These methods are from `SunriseAndSunsetCalculator`, which also adopts `AstronomicalCalculator` protocol. (An Obj-C protocol is like a Java interface in that it defines abstract methods to implement.) 

I think we may want to de-duplicate this code. 

One option is to move it into a common superclass. The cool thing about Obj-C is that classes are just interfaces with implementations. We can technically add an implementation to the Astronomical Calculator class and move the common methods there.

What do you think?

I'd even consider defining an enum which switches the underlying algorithm, but that's for another PR.

Reply:
I like the idea, but I would need to look into how to go about it. We can do that in another PR also.

Reply:
Here's a step-by step way we can do it:

1. Add a new file called `AstronomicalCalculator.m`
2. Add `@implementation...` 
3. Move the implementation into the file 

That might be all we need to do.

Reply:
Yeah, I tried to do it, but I am getting a bunch of errors. I might need help on this issue.

Reply:
Happy to help! What errors are you getting?